server/vulnerabilities/sync.go

```diff
@@ -0,0 +1,263 @@
+package vulnerabilities
+
+import (
+	"context"
+	"encoding/csv"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/facebookincubator/nvdtools/cvefeed"
+	feednvd "github.com/facebookincubator/nvdtools/cvefeed/nvd"
+	"github.com/fleetdm/fleet/v4/pkg/download"
+	"github.com/fleetdm/fleet/v4/pkg/fleethttp"
+	"github.com/fleetdm/fleet/v4/server/config"
+	"github.com/fleetdm/fleet/v4/server/fleet"
+	"github.com/fleetdm/fleet/v4/server/ptr"
+)
+
+// Sync downloads all the vulnerability data sources and loads cve scores into the database.
+func Sync(vulnPath string, config config.FleetConfig, ds fleet.Datastore) error {
+	if config.Vulnerabilities.DisableDataSync {
+		return nil
+	}
+
+	client := fleethttp.NewClient()
+
+	if err := DownloadCPEDatabase(vulnPath, client, WithCPEURL(config.Vulnerabilities.CPEDatabaseURL)); err != nil {
+		return fmt.Errorf("sync CPE database: %w", err)
+	}
+
+	if err := DownloadNVDCVEFeed(vulnPath, ""); err != nil {
+		return fmt.Errorf("sync NVD CVE feed: %w", err)
+	}
+
+	if err := DownloadEPSSFeed(vulnPath, client); err != nil {
+		return fmt.Errorf("sync EPSS CVE feed: %w", err)
+	}
+
+	if err := DownloadCISAKnownExploitsFeed(vulnPath, client); err != nil {
+		return fmt.Errorf("sync CISA known exploits feed: %w", err)
+	}
+
+	if err := LoadCVEScores(vulnPath, ds); err != nil {
+		return fmt.Errorf("load cve scores: %w", err)
+	}
+
+	return nil
+}
+
+const epssFeedsURL = "https://epss.cyentia.com"
+const epssFilename = "epss_scores-current.csv.gz"
+
+// DownloadEPSSFeed downloads the EPSS scores feed.
+func DownloadEPSSFeed(vulnPath string, client *http.Client) error {
+	urlString := epssFeedsURL + "/" + epssFilename
+	u, err := url.Parse(urlString)
+	if err != nil {
+		return fmt.Errorf("parse url: %w", err)
+	}
+	path := filepath.Join(vulnPath, strings.TrimSuffix(epssFilename, ".gz"))
+
+	err = download.DownloadAndExtract(client, u, path)
+	if err != nil {
+		return fmt.Errorf("download %s: %w", u, err)
+	}
+
+	return nil
+}
+
+// epssScore represents the EPSS score for a CVE.
+type epssScore struct {
+	CVE   string
+	Score float64
+}
+
+func parseEPSSScoresFile(path string) ([]epssScore, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	r := csv.NewReader(f)
+	r.Comment = '#'
+	r.FieldsPerRecord = 3
+
+	// skip the header
+	r.Read()
+
+	var epssScores []epssScore
+	for {
+		rec, err := r.Read()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+
+		if len(rec) != 3 {
+			continue
```

Comment:
Please add a comment on why we would ignore a record with `len(rec) != 3`.

Reply:
done